### PR TITLE
initial support for OpenStack

### PR DESCRIPTION
**Pull request by joelddiaz**
- [x] add openstack platform details to API/types
- [x] extend webhook validation
- [x] add deprovisioning support
- [x] add provisioning support
- [x] add hiveutil create-cluster support

Because there is no management of DNS when installing on OpenStack, the DNS setup for the API endpoint for the cluster must be performed in advance. On the OpenStack cluster create or pick a Floating IP that will be used for the installation. For a cluster being installed into `stack.example.com` with cluster name `mycluster`, the DNS A record named `api.mycluster.stack.example.com` should point to the Floating IP created/noted previously.

Also, because there is no management of DNS, after the cluster is installed, create or pick another Floating IP. Make the DNS entry `*.apps.mycluster.stack.example.com` pointing to the Floating IP. Then associate the apps Floating IP with the ingress port created during cluster installation:

```
INGRESS_PORT=$(openstack port list --format value -c Name | awk "/mycluster.*-ingress-port/ {print}")
openstack floating ip set --port ${INGRESS_PORT} FLOATING_IP_FOR_APPS
```

### CODE REVIEW

File: config/crds/hive_v1_clusterdeployment.yaml [hunk 1]
```diff
@@ -226,7 +226,7 @@ spec:
                         will be created.
                       type: string
                   type: object
-                baremetal:
+                bareMetal:
```
Comment (dgoodwin) [inline config/crds/hive_v1_clusterdeployment.yaml]: This looks to be undoing an intentional change, this should be "baremetal"

Comment (joelddiaz) [reply config/crds/hive_v1_clusterdeployment.yaml]: Yep. Must have snuck in with the rebase.

**Comment by gregsheremeta** [comment]
what's the status of this one?

**Comment by dgoodwin** [comment]
I've been trying to test but stuck getting CRC + libvirt up. Hopefully soon.

**Comment by joelddiaz** [comment]
updated PR to catch move from dep to go modules

File: pkg/clusterresource/builder.go [hunk 2]
```diff
@@ -101,6 +101,12 @@ type Builder struct {
 	// ReleaseImage is a specific OpenShift release image to install this cluster with. Will override
 	// ImageSet.
 	ReleaseImage string
+
+	// MachineNetwork is the subnet to use for the cluster's machine network.
+	MachineNetwork string
+
+	// SkipMachinePoolGeneration is set to skip generating MachinePool objects
+	SkipMachinePoolGeneration bool
```
Comment (dgoodwin) [inline pkg/clusterresource/builder.go:109]: What's the use case for this option. I can't immediately think of why we'd want to skip the machine pools.

File: contrib/pkg/utils/openstack/openstack.go [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+	"os"
+	"path/filepath"
+
+	log "github.com/sirupsen/logrus"
+
+	"github.com/openshift/hive/pkg/constants"
+)
+
+// GetCreds reads OpenStack credentials either from either the specified credentials file,
+// ~/.config/openstack/clouds.yaml, or /etc/openstack/clouds.yaml
```
[the remark this answers is not in the document]
Comment (joelddiaz) [reply contrib/pkg/utils/openstack/openstack.go:14]: Yes, I got it from https://docs.openstack.org/python-openstackclient/pike/configuration/index.html#clouds-yaml

File: pkg/apis/hive/v1/openstack/machinepools.go [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+	FlavorName string `json:"type"`
+
+	// RootVolume defines the root volume for instances in the machine pool.
+	// The instances use ephemeral disks if not set.
+	// +optional
+	RootVolume *RootVolume `json:"rootVolume,omitempty"`
+}
+
+// Set sets the values from `required` to `a`.
+func (o *MachinePool) Set(required *MachinePool) {
```
[the remark this answers is not in the document]
Comment (joelddiaz) [reply pkg/apis/hive/v1/openstack/machinepools.go:17]: No. I just copied the file directly from the installer https://github.com/openshift/installer/blob/master/pkg/types/openstack/machinepool.go

File: pkg/clusterresource/builder.go (hunk 2 above)
Comment (joelddiaz) [reply pkg/clusterresource/builder.go:109]: I put it there b/c our current OpenStack support has no support for MachinePools

**Comment by dgoodwin** [comment]
/lgtm

Will iterate if any changes are needed once I'm able to get up and running on OpenStack.

File: contrib/pkg/deprovision/openstack.go [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+)
+
+// openStackOptions is the set of options to deprovision an OpenStack cluster
+type openStackOptions struct {
+	logLevel string
+	infraID  string
+	cloud    string
+}
+
+// NewDeprovisionOpenStackCommand is the entrypoint to create the GCP deprovision subcommand
```
Comment (gregsheremeta) [inline contrib/pkg/deprovision/openstack.go:22]: s/GCP/OpenStack/

**Comment by joelddiaz** [comment]
ok, i sqashed (and grabbed @gregsheremeta 's feedback in there as well)

**Pull request merged**